### PR TITLE
[auto-materialize] Display historic AMP rules

### PR DESCRIPTION
**Pull request by clairelin135**
Currently, the UI uses the current set of automaterialize rules for every past evaluation. This PR modifies the existing behavior by:

- serializing the existent rules as part of the asset evaluation
- loading the serialized rules in the UI, using the current set of rules when rules were not serialized in the evaluation for backcompat

### CODE REVIEW

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16249** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16249" target="_blank">Graphite</a> 
    * **PR #16250** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16250" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16250?utm_source=stack-comment).

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16249** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16249" target="_blank">Graphite</a> 
    * **PR #16254** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16254" target="_blank">Graphite</a> 
    * **PR #16250** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16250" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16250?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 1]
```diff
@@ -500,7 +500,18 @@ def get_auto_materialize_asset_evaluations(
...
+                        neighbor_key
+                    )
+
+                    if auto_materialize_policy is None:
+                        check.failed(f"Expected auto materialize policy on asset {asset_key}")
+
+                    evaluations_by_key[neighbor_key] = evaluation._replace(
+                        asset_key=neighbor_key
+                    )._replace(
+                        rule_snapshots=auto_materialize_policy.rule_snapshots
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:513]: I don't love this pattern, but handling this case gracefully is a bit out of scope for this PR

**Comment by clairelin135** [comment]
@OwenKephart this PR is ready for you to take a look whenever you get the chance

**Review by OwenKephart — APPROVED** [review]
Note the one nit, but otherwise this looks great!
